os_stub/cryptlib_openssl/pk/ec.c

```diff
@@ -846,3 +846,221 @@ bool libspdm_ecdsa_verify(void *ec_context, size_t hash_nid,
+        return 0;
+    }
+
+    if (!EC_KEY_can_sign(eckey)) {
+        return 0;
+    }
+
+    ctx = BN_CTX_new();
+    if (ctx == NULL) {
+        return 0;
+    }
+
+    /* this value is later returned in *kinvp */
+    k = BN_new();
+    /* this value is later returned in *rp */
+    r = BN_new();
+    X = BN_new();
+
+    if (k == NULL || r == NULL || X == NULL) {
+        return 0;
+    }
+    if ((tmp_point = EC_POINT_new(group)) == NULL) {
+        return 0;
+    }
+    order = EC_GROUP_get0_order(group);
+
+    /* Preallocate space */
+    order_bits = BN_num_bits(order);
+    if (!BN_set_bit(k, order_bits)
+        || !BN_set_bit(r, order_bits)
+        || !BN_set_bit(X, order_bits)) {
+        goto err;
+    }
+
+    /*hard code random number*/
+    k = BN_bin2bn(random, random_len, NULL);
+
+    /* compute r the x-coordinate of generator * k */
+    if (!EC_POINT_mul(group, tmp_point, k, NULL, NULL, ctx)) {
+        goto err;
+    }
+    if (!EC_POINT_get_affine_coordinates(group, tmp_point, X, NULL, ctx)) {
+        goto err;
+    }
+    if (!BN_nnmod(r, X, order, ctx)) {
+        goto err;
+    }
+
+    /* compute the inverse of k */
+    if (!ec_group_do_inverse_ord(group, k, k, ctx)) {
+        goto err;
+    }
+
+    /* clear old values if necessary */
+    BN_clear_free(*rp);
+    BN_clear_free(*kinvp);
+    /* save the pre-computed values  */
+    *rp = r;
+    *kinvp = k;
+    ret = 1;
+
+err:
+    if (!ret) {
+        BN_clear_free(k);
+        BN_clear_free(r);
+    }
+
+    BN_CTX_free(ctx);
+    EC_POINT_free(tmp_point);
+    BN_clear_free(X);
+    return ret;
+}
+
+/**
+ * Carries out the EC-DSA signature with caller input random function. This API can be used for FIPS test.
+ *
+ * @param[in]       ec_context    Pointer to EC context for signature generation.
+ * @param[in]       hash_nid      hash NID
+ * @param[in]       message_hash  Pointer to octet message hash to be signed.
+ * @param[in]       hash_size     Size of the message hash in bytes.
+ * @param[out]      signature     Pointer to buffer to receive EC-DSA signature.
+ * @param[in, out]  sig_size      On input, the size of signature buffer in bytes.
+ *                                On output, the size of data returned in signature buffer in bytes.
+ * @param[in]       random_func   random number function
+ *
+ * @retval  true   signature successfully generated in EC-DSA.
+ * @retval  false  signature generation failed.
+ * @retval  false  sig_size is too small.
+ **/
+bool libspdm_ecdsa_sign_ex(void *ec_context, size_t hash_nid,
+                           const uint8_t *message_hash, size_t hash_size,
+                           uint8_t *signature, size_t *sig_size,
+                           int (*random_func)(void *, unsigned char *, size_t))
+{
+    EC_KEY *ec_key;
+    ECDSA_SIG *ecdsa_sig;
+    int32_t openssl_nid;
+    uint8_t half_size;
+    BIGNUM *bn_r;
+    BIGNUM *bn_s;
+    int r_size;
+    int s_size;
+    /* hardcode random number*/
```

Comment:
remove `hardcode random number`

Basic rule: you can hardcode random number in FIPS test, but you cannot hardcode random number in a crypto library.

Reply:
Thanks. I will follow the basic rule.